File: 0022-generate-parentheses/0022-generate-parentheses.cs
using System;
using System.Collections.Generic;

public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        List<string> result = new List<string>();
        GenerateParenthesisHelper(result, "", 0, 0, n);
        return result;
    }

    private void GenerateParenthesisHelper(List<string> result, string current, int open, int close, int n) {
        if (current.Length == n * 2) {
            result.Add(current);
            return;
        }

        if (open < n) {
            GenerateParenthesisHelper(result, current + "(", open + 1, close, n);
        }
        
        if (close < open) {
            GenerateParenthesisHelper(result, current + ")", open, close + 1, n);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.go
func generateParenthesis(n int) []string {
	var result []string
	generateParenthesisHelper(&result, "", 0, 0, n)
	return result
}

func generateParenthesisHelper(result *[]string, current string, open int, close int, n int) {
	if len(current) == n*2 {
		*result = append(*result, current)
		return
	}

	if open < n {
		generateParenthesisHelper(result, current+"(", open+1, close, n)
	}

	if close < open {
		generateParenthesisHelper(result, current+")", open, close+1, n)
	}
}